observability/src/main/java/io/grpc/observability/interceptors/InternalLoggingServerInterceptor.java

```diff
@@ -16,32 +16,224 @@
 
   public interface Factory {
     ServerInterceptor create();
   }
 
   public static class FactoryImpl implements Factory {
+    private final Sink sink;
+    private final LogHelper helper;
+
+    static LogHelper createLogHelper(Sink sink, TimeProvider provider) {
+      return new LogHelper(sink, provider);
+    }
+
+    public FactoryImpl(Sink sink) {
+      this.sink = sink;
+      this.helper = createLogHelper(sink, TimeProvider.SYSTEM_TIME_PROVIDER);
+    }
 
     @Override
     public ServerInterceptor create() {
-      return new InternalLoggingServerInterceptor();
+      return new InternalLoggingServerInterceptor(helper);
+    }
+
+    /**
+     * Closes the sink instance.
+     */
+    public void close() {
+      if (sink != null) {
+        sink.close();
+      }
     }
   }
 
+  private InternalLoggingServerInterceptor(LogHelper helper) {
+    this.helper = helper;
+  }
+
   @Override
   public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(ServerCall<ReqT, RespT> call,
       Metadata headers, ServerCallHandler<ReqT, RespT> next) {
     // TODO(dnvindhya) implement the interceptor
-    return null;
+    final AtomicLong seq = new AtomicLong(1);
+    final String rpcId = UUID.randomUUID().toString();
+    final String authority = call.getAuthority();
+    final String serviceName = call.getMethodDescriptor().getServiceName();
+    final String methodName = call.getMethodDescriptor().getBareMethodName();
+    final SocketAddress peerAddress = LogHelper.getPeerAddress(call.getAttributes());
+    Deadline deadline = Context.current().getDeadline();
+    final Duration timeout = deadline == null ? null
+        : Durations.fromNanos(deadline.timeRemaining(TimeUnit.NANOSECONDS));
+
+    // TODO (dnvindhya): implement isMethodToBeLogged() to check for methods to be logged
+    // according to config. Until then always return true.
+    if (!helper.isMethodToBeLogged(call.getMethodDescriptor().getFullMethodName())) {
+      return next.startCall(call, headers);
+    }
+
+    // Event: EventType.GRPC_CALL_REQUEST_HEADER
+    try {
+      helper.logRequestHeader(
+          seq.getAndIncrement(),
+          serviceName,
+          methodName,
+          authority,
+          timeout,
+          headers,
+          EventLogger.LOGGER_SERVER,
+          rpcId,
+          peerAddress);
+    } catch (Exception e) {
+      // Catching generic exceptions instead of specific ones for all the events.
+      // This way we can catch both expected and unexpected exceptions instead of re-throwing
+      // exceptions to callers which will lead to RPC getting aborted.
+      // Expected exceptions to be caught:
+      // 1. IllegalArgumentException
+      // 2. NullPointerException
+      logger.log(Level.SEVERE, "Unable to log request header", e);
+    }
+
+    ServerCall<ReqT, RespT> wrapperCall =
+        new SimpleForwardingServerCall<ReqT, RespT>(call) {
+          @Override
+          public void sendMessage(RespT message) {
```

Comment:
nit: if the lexical order of methods also reflects the order in which these events take place it will be good. So how about `sendHeaders` and then `sendMessage` ?

Reply:
Updated